src/metrics/structural_complexity.rs

```diff
@@ -0,0 +1,130 @@
+//! calculate structural complexity and warn about overly complex functions
+
+use rustc::lint::*;
+use syntax::codemap::Span;
+use syntax::attr::*;
+use syntax::ast::*;
+use syntax::ptr::P;
+use std::fs::OpenOptions;
+use std::io::Write;
+
+use utils::in_macro;
+
+declare_lint! { pub STRUCTURAL_COMPLEXITY, Warn,
+    "finds functions that should be split up into multiple functions" }
+
+const SC_LIMIT: usize = 25;
+
+pub struct StructuralComplexity;
+
+impl LintPass for StructuralComplexity {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(STRUCTURAL_COMPLEXITY)
+    }
+}
+
+fn compute_cc(cx: &EarlyContext, block: &Block) -> usize {
+    if in_macro(cx, block.span) { return 0; }
+    block.stmts.iter().map(|stmt| match stmt.node {
+        StmtDecl(ref decl, _) => if let DeclLocal(ref loc) = decl.node {
+            loc.init.as_ref().map(|e| compute_cc_expr(cx, &*e)).unwrap_or(0)
+        } else {
+            0
+        },
+        StmtExpr(ref expr, _) => compute_cc_expr(cx, expr),
+        StmtSemi(ref expr, _) => compute_cc_expr(cx, expr),
+        StmtMac(_, _) => unreachable!(),
+    }).sum::<usize>() + compute_cc_multi(cx, block.expr.iter())
+}
+
+fn compute_cc_multi<'a, Iter: Iterator<Item=&'a P<Expr>>>(cx: &EarlyContext, v: Iter) -> usize {
+    v.map(|e| compute_cc_expr(cx, &**e)).sum()
+}
+
+fn compute_cc_expr(cx: &EarlyContext, expr: &Expr) -> usize {
+    if in_macro(cx, expr.span) { return 0; }
+    match expr.node {
+        ExprUnary(_, ref expr)
+        | ExprCast(ref expr, _)
+        | ExprParen(ref expr)
+        | ExprField(ref expr, _)
+        | ExprTupField(ref expr, _)
+        | ExprAddrOf(_, ref expr)
+        | ExprBox(ref expr) => compute_cc_expr(cx, &**expr),
+        ExprTup(ref v)
+        | ExprMethodCall(_, _, ref v)
+        | ExprVec(ref v) => compute_cc_multi(cx, v.iter()),
+        ExprCall(ref fun, ref args) => compute_cc_expr(cx, &**fun) + compute_cc_multi(cx, args.iter()),
+        ExprAssign(ref lhs, ref rhs)
+        | ExprAssignOp(_, ref lhs, ref rhs)
+        | ExprIndex(ref lhs, ref rhs)
+        | ExprRepeat(ref lhs, ref rhs)
+        | ExprInPlace(ref lhs, ref rhs)
+        | ExprBinary(_, ref lhs, ref rhs) => compute_cc_expr(cx, &*lhs) + compute_cc_expr(cx, &*rhs),
+        ExprLit(_)
+        | ExprInlineAsm(_)
+        | ExprPath(..) => 0,
+        ExprIfLet(_, ref cond, ref body, ref els)
+        | ExprIf(ref cond, ref body, ref els) =>
+            compute_cc_expr(cx, &**cond) + compute_cc(cx, &**body) + compute_cc_multi(cx, els.iter()) + 1,
+        ExprWhileLet(_, ref cond, ref body, _)
+        | ExprForLoop(_, ref cond, ref body, _)
+        | ExprWhile(ref cond, ref body, _) => compute_cc_expr(cx, &**cond) + compute_cc(cx, &**body) + 1,
+        ExprLoop(ref block, _) => compute_cc(cx, &**block),
+        ExprMatch(ref expr, ref arms) => compute_cc_expr(cx, &**expr) + arms.iter().map(|arm|
```

Comment:
There's also no reason to. Here I made sure that a large match will only increase the complexity by one, plus the complexity in its arms. So if you want to reduce the complexity, you can simply call a function in the arm instead of putting the code in there directly.